pkg/pipeline/encode/encode_prom.go

```diff
@@ -122,34 +133,116 @@ func (e *encodeProm) EncodeGauge(metric config.GenericMap) []interface{} {
-			gaugeName:  e.prefix + gaugeName,
-			gaugeValue: valueFloat,
-			labels:     entryLabels,
+		entry := entryInfo{
+			eInfo: entrySignature{
+				Name:   e.prefix + gaugeName,
+				Labels: entryLabels,
+			},
+			value: valueFloat,
 		}
 		out = append(out, entry)
 		// push the metric to prometheus
-		// TODO - fix the types here
 		if gaugeInfo.promGauge != nil {
 			gaugeInfo.promGauge.With(entryLabels).Set(valueFloat)
 		}
-	}
 
+		cEntry := e.saveEntryInCache(entry, entryLabels)
+		cEntry.metricType = api.PromEncodeOperationName("Gauge")
+		cEntry.promGauge = gaugeInfo.promGauge
+	}
 	return out
 }
 
+func (e *encodeProm) saveEntryInCache(entry entryInfo, entryLabels map[string]string) *metricCacheEntry {
+	// save item in cache; use eInfo as key to the cache
+	var cEntry *metricCacheEntry
+	secs := time.Now().Unix()
+	eInfoBytes, _ := json.Marshal(&entry.eInfo)
+	eInfoString := string(eInfoBytes)
+	cEntry, ok := e.mCache[eInfoString]
+	if ok {
+		// item already exists in cache; update the element and move to end of list
+		cEntry.timeStamp = secs
+		// move to end of list
+		e.mList.MoveToBack(cEntry.e)
+	} else {
+		// create new entry for cache
+		cEntry = &metricCacheEntry{
+			label:     entryLabels,
+			timeStamp: secs,
+			key:       eInfoString,
+		}
+		// place at end of list
+		log.Debugf("adding entry = %v", cEntry)
+		cEntry.e = e.mList.PushBack(cEntry)
+		e.mCache[eInfoString] = cEntry
+		log.Debugf("mlist = %v", e.mList)
+	}
+	return cEntry
+}
+
 // Encode encodes a flow before being stored
 func (e *encodeProm) Encode(metrics []config.GenericMap) []interface{} {
+	log.Debugf("entering encodeProm Encode")
+	e.mu.Lock()
+	defer e.mu.Unlock()
 	out := make([]interface{}, 0)
 	for _, metric := range metrics {
 		gaugeOut := e.EncodeGauge(metric)
 		out = append(out, gaugeOut...)
 		counterOut := e.EncodeCounter(metric)
 		out = append(out, counterOut...)
 	}
+	log.Debugf("cache = %v", e.mCache)
+	log.Debugf("list = %v", e.mList)
 	return out
 }
 
+func (e *encodeProm) cleanupExpiredEntriesLoop() {
+	for {
+		e.cleanupExpiredEntries()
+		time.Sleep(time.Duration(e.expiryTime) * time.Second)
+	}
+}
+
+// cleanupExpiredEntries - any entry that has expired should be removed from the prometheus reporting and cache
+func (e *encodeProm) cleanupExpiredEntries() {
+	log.Debugf("entering cleanupExpiredEntries")
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	log.Debugf("cache = %v", e.mCache)
+	log.Debugf("list = %v", e.mList)
+	secs := time.Now().Unix()
+	expireTime := secs - e.expiryTime
+	// go through the list until we reach recently used connections
+	for {
+		entry := e.mList.Front()
+		if entry == nil {
+			return
+		}
+		c := entry.Value.(*metricCacheEntry)
+		log.Debugf("timeStamp = %d, expireTime = %d", c.timeStamp, expireTime)
+		log.Debugf("c = %v", c)
+		if c.timeStamp > expireTime {
+			// no more expired items
+			return
+		}
+
+		// clean up the entry
+		log.Debugf("secs = %d, deleting %s", secs, c.label)
```

Comment:
you probably want to add also expireTime to the debug print here

Reply:
expireTime appears in a separate debug print just above.